### ISSUE TITLE
API authentication/CSRF protection

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#39]
Addresses https://github.com/PrivateBin/PrivateBin/issues/20 - 2.10

Options:
- API key in URL (random string)
- Some auth method with CORS headers
- CORS - only same domain instead of `*`
- OAUTH (?)

Depends on use-case of API. E.g. when the API is used by desktop/mobile applications CORS headers are irrelevant and the desktop/mobile application is assumed to be handle all things securely.
CORS headers are only relevant for applications on websites using PrivateBin API. For this use-case we should look at similar APIs, e.g. how GitHub does it.

**Comment by rugk** [comment]
I would say the API token would be the easiest and best method where not much change is needed.

The issue is only for that the token should be valid:
* Server token would allow to manage all pads  
  Pro: Would prevent easy CSRF attacks  
  Con: Compromise of API key is very bad and to spread it again is difficult
* Per user  
  Con: Would need some way to authenticate.
* Per paste/bin  
  Pro: API keys can be managed separately.  
  Con: A new API key for each paste is horrible UX. Still needs server key for paste generation.

This are basically my ideas. No one looks really good, but I think the server key is still the best option.

**Comment by elrido** [comment]
I would go for an option to enable or disable CORS for all (*) or current server only. API tokens would not be logical as they would require authentication (user accounts) on each server involved.

The idea of the API was to explicitly allow local PrivateBin installs or Apps to use another (third party) server as storage. This means that the current CSRF is by design. And we do not want to introduce user accounts or authentication besides of the current optional password.

This is a design challenge. I would strongly suggest to postpone this for now and address it as part of #2. Any solution we would do up front might make it impossible to introduce the paste manager feature without breaking the API again. Lets do it only once and then properly.

**Comment by rugk** [comment]
> API tokens would not be logical as they would require authentication (user accounts) on each server involved.

As said, not necessarily. What about a server API keys you can get with a few clicks?
Ah just saw why this is stupid: Anyone can get this API key, also an attacker could use it...

> I would go for an option to enable or disable CORS for all (*) or current server only. 

This would certainly be the easiest solution.
And I think this is also the best solution at the moment.
